### ISSUE TITLE
Looping in parallel tasks

### ISSUE THREAD

**Issue by kamil-zbroniec** [danielgerlag/workflow-core#372]
Hi,

I am trying to create a workflow, that will consist of two parts - first responsible for keeping workflow configuration up-to-date (it waits for config updated event and reloads the configuration from DB) and second responsible for doing the actual job.
I started experimenting with parallel tasks as I want at some point to be able to wait for two different events at the same time and came up with something like this:
```
builder
    .StartWith<GetConfigurationFromDB>() // initial
    .Parallel()
        .Do(configWatch =>
        {
            configWatch
                .StartWith<Blank>(config => // Blank is { return ExecutionResult.Next(); }
                {
                    config
                        .WaitFor("config_updated", data => data.ConfigUpdatedEventKey, data => DateTime.UtcNow)
                        .Then<GetConfigurationFromDB>() // onUpdate
                        .Then(config);
                });
        })
        .Do(actualJob =>
        {
            actualJob
                .StartWith<DoSomething>()
                .Then<DoSomethingElse>(doSomething => 
                        .WaitFor("some_event", data => data.SomeEventKey, data => DateTime.UtcNow)
                .Then(doSomething));
        })
    .Join();
```
However the outcome of the above code is like this:
```
/ Workflow starts /
GetConfigurationFromDB (initial)
GetConfigurationFromDB (intial) - again for some reason
/ config_updated event is raised /
GetConfigurationFromDB (onUpdate)
GetConfigurationFromDB (initial) - again
/ Another config_updated event is raised /
GetConfigurationFromDB (onUpdate)
GetConfigurationFromDB (initial)
GetConfigurationFromDB (onUpdate)
GetConfigurationFromDB (initial)
```
If I modify the code in the following way, the issue stops and the execution is done properly:
```
builder
    .StartWith<GetConfigurationFromDB>() // initial
    .Parallel()
        .Do(configWatch =>
        {
            configWatch
                .StartWith<Blank>() // Blank is { return ExecutionResult.Next(); }
                .Then<Blank>(config => 
                {
                    config
                        .WaitFor("config_updated", data => data.ConfigUpdatedEventKey, data => DateTime.UtcNow)
                        .Then<GetConfigurationFromDB>() // onUpdate
                        .Then(config);
                });
        })
        .Do(actualJob =>
        {
            actualJob
                .StartWith<DoSomething>()
                .Then<DoSomethingElse>(doSomething => 
                        .WaitFor("some_event", data => data.SomeEventKey, data => DateTime.UtcNow)
                .Then(doSomething));
        })
    .Join();
```
```
/ Workflow starts /
GetConfigurationFromDB (initial)
/ config_updated event is raised /
GetConfigurationFromDB (onUpdate)
/ Another config_updated event is raised /
GetConfigurationFromDB (onUpdate)
```
The question is why? What is the proper way of using loops in this situation? Is the second approach recommended? I use WorkflowCore version 2.0.0.

Many thanks for the answer.

**Comment by danielgerlag** [comment]
Did you have more than 1 workflow instance active the first time?

**Comment by kamil-zbroniec** [comment]
No. Each time I had one workflow instance.

W dniu niedz., 25.08.2019 o 03:25 Daniel Gerlag <notifications@github.com>
napisał(a):

**Comment by danielgerlag** [comment]
Your syntax looks odd, I'm not sure why you are building a sequence off of `Blank`... please have a look at the docs and samples for Parallel
https://workflow-core.readthedocs.io/en/latest/control-structures/#parallel-paths
https://github.com/danielgerlag/workflow-core/tree/master/src/samples/WorkflowCore.Sample13

**Comment by kamil-zbroniec** [comment]
I agree it looks odd :) That's because I want to include Looping, Parallel and Events in one workflow. It would be perfect if the syntax could look like that:
```
.Do(configWatch =>
{
    configWatch
        .WaitFor("config_updated", data => data.ConfigUpdatedEventKey, data => DateTime.UtcNow)
        .Then<GetConfigurationFromDB>() // onUpdate
        .Then(configWatch);
})
```
But it can't because of incompatible types (`configWatch` is of type `IWorkflowBuilder`, `WaitFor` is of type `IStepBuilder`, `Then` takes `IStepBuilder` as an argument). That is why I started to experiment with putting `StartWith` and `Then` inside `Do` and as there was really nothing I'd like to be done in these steps I decided to create and empty step called `Blank`.

**Comment by danielgerlag** [comment]
Why don't you just use a `While` loop?

**Comment by kamil-zbroniec** [comment]
I can do something like that:
```
.Do(configWatch =>
{
    configWatch
        .StartWith<Blank>()
        .While(x => true)
            .Do(then =>
            {
                then
                    .StartWith<Blank>()
                    .WaitFor("config_updated", data => data.ConfigUpdatedEventKey, data => DateTime.UtcNow)
                    .Then<GetConfigurationFromDB>()
            });
})
```
But I am not sure if it is in fact an improvement to what I've already done. 

Nevertheless I think you're missing a point of my initial question. I was wondering why is there this odd behaviour while looping inside `StartWith` method, for example:
```
.Do(configWatch =>
{
    configWatch
        .StartWith<Blank>(config =>
        {
            config
                .WaitFor("config_updated", data => data.ConfigUpdatedEventKey, data => DateTime.UtcNow)
                .Then<GetConfigurationFromDB>()
                .Then(config);
        });
})
```
Is it because of some bug in the code or is it because of my misuse of your code? Or maybe both?
